junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt

```diff
@@ -257,8 +439,13 @@ fun <R> assertTimeout(timeout: Duration, message: () -> String, executable: () -
  * @paramR the result of the [executable].
  */
 @API(status = EXPERIMENTAL, since = "5.5")
-fun <R> assertTimeoutPreemptively(timeout: Duration, executable: () -> R): R =
-    Assertions.assertTimeoutPreemptively(timeout, executable)
+fun <R> assertTimeoutPreemptively(timeout: Duration, executable: () -> R): R {
+    contract {
+        callsInPlace(executable, EXACTLY_ONCE)
```

Comment:
Is it actually true for things that can/expected to throw?
(I have a feeling we can't say "exactly once", but I might be misinterpreting the enum.)
Examples where it would not have compiled without the contract, and now it's a runtime error:

```kotlin
val value: String
// Swapped version of fun `assertThrows with value initialization in lambda`() in this PR
assertThrows<AssertionError> {
    Assertions.fail("message")
    value = "string"
}
assertEquals("string", value) // AssertionFailedError: expected: <string> but was: <null>
```

```kotlin
val x: String
assertThrows<IOException> {
    File("//").readText() // Deterministic IOException due to invalid name.
    x = "abc" // Will never execute, because it always throws.
}
// Kotlin thinks x is initialized here, because of the EXACTLY_ONCE.
assertEquals(3, x.length) // Consistent NPE, due to "effectively unreachable code" above.
```

```kotlin
val x: String
assertThrows<AssertionError> { // EXACTLY_ONCE
    assertTimeoutPreemptively(Duration.seconds(1)) { // EXACTLY_ONCE
        Thread.sleep(2000)
        x = "" // Will never execute, because it always times out.
    }
}
// Kotlin thinks x is initialized here, because of the chain of EXACTLY_ONCE.
println(x.length) // Consistent NPE, due to "effectively unreachable code" above.
```

Reply:
What happens if you write this in Kotlin with try/catch? I presume the compiler pitches a fit?

```kotlin
val x: String
try {
    File("//").readText() // Deterministic IOException due to invalid name.
    x = "abc" // Will never execute, because it always throws.
} catch (e: IOException) {
    
}
assertEquals(3, x.length)
```

What about this?

```kotlin
val x: String
File("//").apply {
	readText() // Deterministic IOException due to invalid name.
    x = "abc" // Will never execute, because it always throws.
}

assertEquals(3, x.length)
```

If the `apply` case fails in the same way as your example, I think it's probably fine?

Reply:
This looks like it will compile fine:

```
import java.io.File

var x: String
File("//").apply {
    readText() // Deterministic IOException due to invalid name.
    x = "abc" // Will never execute, because it always throws.
}

if (3 == x.length) {
    
}
```

Reply:
Huh, very good point! 😁

try-catch example will not compile, but because not all code paths initialize x, but the point is valid there too, you can resolve this thread (I can't).

Reply:
You could theoretically reimplement this logic, in kotlin, and expose it as an inline function. Then the try/catch will be inlined. Then there will be no need for the contract

https://github.com/junit-team/junit5/blob/ffab3a37f4a4e33529cca4ef9a464c244756c6fc/junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrows.java#L49-L74

Reply:
Contracts don't work well when we suppress exception. Issues [one](https://youtrack.jetbrains.com/issue/KT-31692) and [two](https://youtrack.jetbrains.com/issue/KT-43092).
To prevent any kind of errors and confusion, I suggest to replace `callsInPlace(executable, EXACTLY_ONCE)` with `callsInPlace(executable)` in all `assertThrows` methods.

What do you think?

Reply:
I haven't noticed any improvements in contracts. `InvocationKind.UNKNOWN` is set for `assertTimeoutPreemptively` methods because of the same reasons as for `assertThrows` methods